module/os/linux/zfs/zpl_file.c

```diff
@@ -998,6 +998,129 @@ zpl_ioctl_setxattr(struct file *filp, void __user *arg)
 	return (err);
 }
 
+/*
+ * Expose Additional File Level Attributes of ZFS.
+ */
+static int
+zpl_ioctl_getdosflags(struct file *filp, void __user *arg)
+{
+	struct inode *ip = file_inode(filp);
+	uint64_t dosflags = ITOZ(ip)->z_pflags;
+	dosflags &= ZFS_KEEP_ADDITIONAL_FLAGS;
+	int err = copy_to_user(arg, &dosflags, sizeof (dosflags));
+
+	return (err);
+}
+
+static int
+__zpl_ioctl_setdosflags(struct inode *ip, uint64_t ioctl_flags, xvattr_t *xva)
+{
+	uint64_t zfs_flags = ITOZ(ip)->z_pflags;
+	xoptattr_t *xoap;
+
+	if (ioctl_flags & ~(ZFS_IMMUTABLE | ZFS_APPENDONLY | ZFS_NOUNLINK |
+	    ZFS_ARCHIVE | ZFS_NODUMP | ZFS_SYSTEM | ZFS_HIDDEN |
+	    ZFS_READONLY | ZFS_REPARSE | ZFS_OFFLINE | ZFS_SPARSE))
+		return (-EOPNOTSUPP);
```

Comment:
It looks like `ZFS_OPAQUE`, `ZFS_AV_QUARANTINED`, and `ZFS_AV_MODIFIED` were not included here.  Could you comment on why there weren't included?

Reply:
@behlendorf From `zfs_freebsd_setattr()`, it looks like `EOPNOTSUPP` is returned for `ZFS_OPAQUE`, `ZFS_AV_MODIFIED` and `ZFS_AV_QUARANTINED`. If you recommend adding those here for Linux, I can add them.

Reply:
Got it.  Then let's keep it consistent with FreeBSD, when skimming `zfs_freebsd_setattr()` I missed these returned `EOPNOTSUPP` since they used the `SF_` / `UF_` names.